plugins/chain_plugin/chain_plugin.cpp

```diff
@@ -934,6 +934,11 @@ void chain_plugin_impl::plugin_initialize(const variables_map& options) {
 
       chain_config->db_map_mode = options.at("database-map-mode").as<pinnable_mapped_file::map_mode>();
 
+      // when loading a snapshot, all the state will be modified, so use the `shared` mode instead
+      // of `copy_on_write` to lower memory requirements
+     if (snapshot_path && chain_config->db_map_mode == pinnable_mapped_file::mapped)
+        chain_config->db_map_mode = pinnable_mapped_file::mapped_shared;
```

Comment:
> It would be nice to maintain the new `mapped` mode for loading snapshots because I think it would be a huge perf boost (I've seen comments from some users it takes 30 minutes to load a WAX snapshot, but it takes me less than 5 minutes in `heap` mode.. I'm pretty sure it's disk grinding for those users).

This would be useful when you have just the right amount of RAM that can hold all the state in RAM, but not quite enough for `heap` mode. I'm a little bit concerned that we may get more crashes this way. Maybe I can detect the available RAM, and according to the size of the disk db configured decide if it makes sense to use the new `mapped` mode (for example go for it if RAM_size > 1.1 x db_size).

> If we're worried about leaving 100% dirty pages after loading a snapshot, maybe one option is chainbase could expose a flush() call that (synchronously) performs the write out so that all the pages are clean. and nodeos calls that after loading the snapshot but before continuing.

Yes that's a great idea. We can still use the soft-dirty thing as is the state-db-size is still configured much greater than the actual db_size used, it will make the flush faster.

Reply:
Should add an `ilog` or maybe even a `wlog` that the mode was changed from specified.

Reply:
Upon reflection, I think the best compromise when loading a snapshot is:
- load the snapshot in `mapped_shared` mode. Yes it is probably slower, but it minimizes the odds of running out of memory. With the new `mapped` mode, we need memory for both the currently read data from the snapshot + the full chainbase db.
- when the snapshot is done loading, use a new API as suggested by Matt which flushes the still dirty pages to disk and restart with a new `copy_on_write` mapping.

Do you guys agree?

Reply:
I just checked in the implementation of the above.

Reply:
@heifner I updated the change to `mapped_shared` to be temporary (just while loading the snapshot), so I don't think a `ilog` is necessary, but I can add it if you think it might be useful still.

Reply:
No, not needed if it honors the configuration after snapshot load.